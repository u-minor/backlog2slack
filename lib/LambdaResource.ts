import { Duration, Stack } from 'aws-cdk-lib';
import {
  FunctionUrl,
  FunctionUrlAuthType,
  Runtime,
} from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';

export interface LambdaResourceProps {
  backlog: {
    apiKey: string;
    host: string;
  };
  slack: {
    botToken: string;
    signingSecret: string;
  };
}

export class LambdaResource {
  functionUrl: FunctionUrl;
  nodejsFunction: NodejsFunction;

  constructor(stack: Stack, props: LambdaResourceProps) {
    this.nodejsFunction = new NodejsFunction(stack, 'Lambda', {
      description: 'Send Backlog notification to Slack',
      entry: 'src/index.ts',
      environment: {
        BACKLOG_API_KEY: props.backlog.apiKey,
        BACKLOG_HOST: props.backlog.host,
        SLACK_BOT_TOKEN: props.slack.botToken,
        SLACK_SIGNING_SECRET: props.slack.signingSecret,
      },
      functionName: `${stack.stackName}-Lambda`,
      handler: 'handler',
      logRetention: 7,
      memorySize: 128,
      runtime: Runtime.NODEJS_18_X,
      timeout: Duration.seconds(30),
    });

    this.functionUrl = this.nodejsFunction.addFunctionUrl({
      authType: FunctionUrlAuthType.NONE,
      cors: {
        allowedHeaders: ['*'],
        allowedOrigins: ['*'],
        maxAge: Duration.seconds(300),
      },
    });
  }
}
