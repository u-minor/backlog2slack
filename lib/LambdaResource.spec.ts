import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { LambdaResource } from './LambdaResource';

describe('LambdaResource', () => {
  let template: Template;

  beforeAll(() => {
    const app = new App();
    const stack = new Stack(app, 'TestStack');

    new LambdaResource(stack, {
      backlog: {
        apiKey: 'TEST_BACKLOG_API_KEY',
        host: 'example.backlog.jp',
      },
      slack: {
        botToken: 'TEST_SLACK_BOT_TOKEN',
        signingSecret: 'TEST_SLACK_SIGNING_SECRET',
      },
    });

    template = Template.fromStack(stack);
  });

  it('should generate valid IAM role resources', () => {
    const actual = template.findResources('AWS::IAM::Role', {
      Properties: {
        AssumeRolePolicyDocument: {
          Statement: [{ Principal: { Service: 'lambda.amazonaws.com' } }],
        },
      },
    });

    expect(actual).toMatchSnapshot();
  });

  it('should generate valid Lambda function resources', () => {
    const actual = template.findResources('AWS::Lambda::Function');

    expect(actual).toMatchSnapshot();
  });

  it('should generate valid Lambda function url resource', () => {
    const actual = template.findResources('AWS::Lambda::Url');

    expect(actual).toMatchSnapshot();
  });
});
