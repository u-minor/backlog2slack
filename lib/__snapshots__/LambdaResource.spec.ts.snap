// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LambdaResource should generate valid IAM role resources 1`] = `
{
  "LambdaServiceRoleA8ED4D3B": {
    "Properties": {
      "AssumeRolePolicyDocument": {
        "Statement": [
          {
            "Action": "sts:AssumeRole",
            "Effect": "Allow",
            "Principal": {
              "Service": "lambda.amazonaws.com",
            },
          },
        ],
        "Version": "2012-10-17",
      },
      "ManagedPolicyArns": [
        {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
            ],
          ],
        },
      ],
    },
    "Type": "AWS::IAM::Role",
  },
  "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
    "Properties": {
      "AssumeRolePolicyDocument": {
        "Statement": [
          {
            "Action": "sts:AssumeRole",
            "Effect": "Allow",
            "Principal": {
              "Service": "lambda.amazonaws.com",
            },
          },
        ],
        "Version": "2012-10-17",
      },
      "ManagedPolicyArns": [
        {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
            ],
          ],
        },
      ],
    },
    "Type": "AWS::IAM::Role",
  },
}
`;

exports[`LambdaResource should generate valid Lambda function resources 1`] = `
{
  "LambdaD247545B": {
    "DependsOn": [
      "LambdaServiceRoleA8ED4D3B",
    ],
    "Properties": {
      "Code": {
        "S3Bucket": {
          "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
        },
        "S3Key": "419d11c84608d86b57fd372a92c7da56bbf630c4e14c25d2fd595dbad7ca6cff.zip",
      },
      "Description": "Send Backlog notification to Slack",
      "Environment": {
        "Variables": {
          "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          "BACKLOG_API_KEY": "TEST_BACKLOG_API_KEY",
          "BACKLOG_HOST": "example.backlog.jp",
          "SLACK_BOT_TOKEN": "TEST_SLACK_BOT_TOKEN",
          "SLACK_SIGNING_SECRET": "TEST_SLACK_SIGNING_SECRET",
        },
      },
      "FunctionName": "TestStack-Lambda",
      "Handler": "index.handler",
      "MemorySize": 128,
      "Role": {
        "Fn::GetAtt": [
          "LambdaServiceRoleA8ED4D3B",
          "Arn",
        ],
      },
      "Runtime": "nodejs18.x",
      "Timeout": 30,
    },
    "Type": "AWS::Lambda::Function",
  },
  "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
    "DependsOn": [
      "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
      "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
    ],
    "Properties": {
      "Code": {
        "S3Bucket": {
          "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
        },
        "S3Key": "5bc602ecde93c947efe5899ae355f999986a1acbe610b1c0b9c468d738857555.zip",
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          "Arn",
        ],
      },
      "Runtime": "nodejs18.x",
      "Timeout": 900,
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`LambdaResource should generate valid Lambda function url resource 1`] = `
{
  "LambdaFunctionUrl47093E15": {
    "Properties": {
      "AuthType": "NONE",
      "Cors": {
        "AllowHeaders": [
          "*",
        ],
        "AllowMethods": [
          "*",
        ],
        "AllowOrigins": [
          "*",
        ],
        "MaxAge": 300,
      },
      "TargetFunctionArn": {
        "Fn::GetAtt": [
          "LambdaD247545B",
          "Arn",
        ],
      },
    },
    "Type": "AWS::Lambda::Url",
  },
}
`;
